src/main/java/bio/terra/service/resourcemanagement/ResourceService.java

```diff
@@ -36,67 +37,98 @@
     public ResourceService(
         DataLocationSelector dataLocationSelector,
         GoogleProjectService projectService,
         GoogleBucketService bucketService,
-        SamConfiguration samConfiguration) {
+        SamConfiguration samConfiguration,
+        DatasetBucketDao datasetBucketDao,
+        GoogleResourceDao resourceDao) {
         this.dataLocationSelector = dataLocationSelector;
         this.projectService = projectService;
         this.bucketService = bucketService;
         this.samConfiguration = samConfiguration;
+        this.datasetBucketDao = datasetBucketDao;
+        this.resourceDao = resourceDao;
     }
 
     /**
      * Fetch/create a project, then use that to fetch/create a bucket.
      *
-     * @param billingProfile authorized profile for billing account information case we need to create a project
+     * @param billingProfile authorized profile for billing account information case we need to
+     *                       create a project
      * @param flightId       used to lock the bucket metadata during possible creation
      * @return a reference to the bucket as a POJO GoogleBucketResource
      * @throws CorruptMetadataException in two cases.
-     * <ol>
-     *     <le>if the bucket already exists, but the metadata does not AND
-     *     the application property allowReuseExistingBuckets=false.</le>
-     *     <le>if the metadata exists, but the bucket does not</le>
-     * </ol>
+     *                                  <ol>
+     *                                      <le>if the bucket already exists, but the metadata does not AND
+     *                                      the application property allowReuseExistingBuckets=false.</le>
+     *                                      <le>if the metadata exists, but the bucket does not</le>
+     *                                  </ol>
      */
-    public GoogleBucketResource getOrCreateBucketForFile(Dataset dataset,
-                                                         BillingProfileModel billingProfile,
-                                                         String flightId) throws InterruptedException {
+    public GoogleBucketResource getOrCreateBucketForFile(String possibleBucketName,
+        Dataset dataset,
+        BillingProfileModel billingProfile,
+        String flightId) throws InterruptedException {
         final String datasetName = dataset.getName();
         // Every bucket needs to live in a project, so we get or create a project first
         final GoogleProjectResource projectResource = projectService.getOrCreateProject(
             dataLocationSelector.projectIdForFile(datasetName, billingProfile),
             billingProfile,
             null);
 
-        return bucketService.getOrCreateBucket(
-            dataLocationSelector.bucketForFile(datasetName, billingProfile),
-            projectResource,
+        List<UUID> bucketsForDataset = datasetBucketDao.getBucketForDatasetId(dataset.getId());
```

Comment:
RE my previous suggestion - this chunk of code (lines 85-95) could be pulled into their own method and called in the "BucketNameStep"